src/Components/Components/src/Virtualization/VirtualizeBase.cs

```diff
@@ -0,0 +1,128 @@
+
+        /// <summary>
+        /// Gets or sets the item template for the list.
+        /// </summary>
+        [Parameter]
+        public RenderFragment<TItem>? ChildContent { get; set; }
+
+        /// <summary>
+        /// Gets or sets the item template for the list.
+        /// </summary>
+        [Parameter]
+        public RenderFragment<TItem>? Item { get; set; }
+
+        /// <summary>
+        /// Gets or sets the template of the footer to be rendered at the end of the list.
+        /// </summary>
+        [Parameter]
+        public RenderFragment? Footer { get; set; }
+
+        /// <summary>
+        /// Gets the size of each item in pixels.
+        /// </summary>
+        [Parameter]
+        public float ItemSize { get; set; }
+
+        /// <summary>
+        /// Gets or sets the fixed item source.
+        /// </summary>
+        [Parameter]
+        public ICollection<TItem>? Items { get; set; }
+
+        /// <summary>
+        /// Gets or sets the function retrieving items given a start index.
+        /// </summary>
+        [Parameter]
+        public Func<int, Task<IEnumerable<TItem>>>? ItemsProvider { get; set; }
+
+        /// <inheritdoc />
+        public override async Task SetParametersAsync(ParameterView parameters)
+        {
+            await base.SetParametersAsync(parameters);
+
+            if (Items != null)
+            {
+                if (ItemsProvider != null)
+                {
+                    throw new InvalidOperationException($"{GetType()} cannot accept multiple item sources.");
+                }
+
+                ItemSource = Items;
+            }
+            else if (ItemsProvider != null)
+            {
+                ItemSource = _loadedItems;
+            }
+            else
+            {
+                throw new InvalidOperationException($"{GetType()} requires either '{nameof(Items)}' or '{nameof(ItemsProvider)}' to be specified and non-null.");
+            }
+
+            ItemTemplate = Item ?? ChildContent;
+        }
+
+        /// <summary>
+        /// Fetches items using <see cref="ItemsProvider"/>.
+        /// If <see cref="ItemsProvider"/> is <c>null</c> or items are already being fetched, this method
+        /// performs nothing.
+        /// </summary>
+        /// <returns>The <see cref="Task"/> representing the completion of this operation.</returns>
+        protected async Task FetchItemsAsync()
+        {
+            if (ItemsProvider == null)
+            {
+                return;
+            }
+
+            if (Interlocked.Exchange(ref _isFetchingItems, 1) == 0)
+            {
+                // TODO: Handle exceptions from the items provider
+                var items = await ItemsProvider(_loadedItems.Count);
+
+                foreach (var item in items)
+                {
+                    _loadedItems.Enqueue(item);
+                }
+
+                _isFetchingItems = 0;
+
+                StateHasChanged();
```

Comment:
Since the only call to `FetchItemsAsync` is followed by a `StateHasChanged`, you probably don't need the `StateHasChanged` call here (assuming the caller will `await` the returned task).